drivers: rtc: sam0: Optimize RTC register access

sslupsky opened the pull request:
The driver issues a READREQ for every access to the COUNT register which causes 180us of delay.  This is a large delay that causes several problems with other systems including i2c and power management.

This commit uses continuous read requests (RCONT) to optimize accessing the COUNT register.  This reduces (but does not eliminate) the time spent synchronizing the COUNT register.

This commit also prevent interrupts from corrupting COUNT and COMP register access.

The kernel does not like it when the RTC does not tick after waking from sleep. RCONT enables continuously syncing the COUNT register but after sleep, we need to wait for a new read request to complete. However, the SYNCBUSY flag is not set when RCONT is enabled so we cannot use that to do the sync. So, we wait for the RTC COUNT register to begin ticking again by reading the COUNT register.

We should wait a minimum of the TICK_THRESHOLD which is the amount of time it takes to sync the register.

Fixes #21549
Fixes #21114
Fixes #21092

sslupsky committed:
drivers: sam0_rtc_timer: Optimize RTC register access

The driver issues a READREQ for every access to the COUNT
register which causes 180us of delay.  This is a large
delay that causes several problems with other systems
including i2c.

This commit uses continuous read requests (RCONT) to
optimize accessing the COUNT register.  This reduces (but does not
eliminate) the time spent synchronizing the COUNT register.
This commit also prevent interrupts from corrupting COUNT and COMP
register access.

The kernel does not like it when the RTC does not tick
after waking from sleep.  RCONT enables continuously syncing the
COUNT register but after sleep, we need to wait for a new read
request to complete.  However, the SYNCBUSY flag is not set
when RCONT is enabled so we cannot use that to do the sync.
So, we wait for the RTC COUNT register to begin ticking again by
reading the COUNT register.

We should wait a minimum of the TICK_THRESHOLD which is the amount
of time it takes to sync the register.

Fixes #21549, #21114, #21092

Fix comment typo

The worst case update time after sleep requires 2 sync busy
synchronizations.

This commit updates the TICK_THRESHOLD to reflect the
additional time required for the additional sync busy.

Fix clock initialisation conflict.
Fix whitespace

mnkp recommended that a separate commit be created for the
system power management.  This commit separates the
power_samd2x.c into a separate commit.

Signed-off-by: Steven Slupsky <sslupsky@gmail.com>

sslupsky commented on drivers/timer/sam0_rtc_timer.c:
```diff
@@ -29,7 +33,7 @@
 #endif
 
 /* Number of sys timer cycles per on tick. */
```
fixed

sslupsky commented:
I would have handled the merge conflict but there seems to be a problem and I cannot build my application with the conflicting change from master.  I think there might be something wrong with the new dts api implementation as the GCLK_EVAL macro does appear to generate the correct reference in the device tree.

Here is the error:
```
/Users/stevenslupsky/Documents/source/zephyrproject/zephyr/drivers/timer/sam0_rtc_timer.c: In function 'z_clock_driver_init':
/Users/stevenslupsky/Documents/source/zephyrproject/zephyr/drivers/timer/sam0_rtc_timer.c:72:22: error: 'GCLK_CLKCTRL_GEN_GCLKDT_N_S_soc_S_rtc_40001400_P_clocks_IDX_0_VAL_generator' undeclared (first use in this function); did you mean 'DT_N_S_soc_S_rtc_40001400_P_clock_generator'?
 #define GCLK_EVAL(n) GCLK_CLKCTRL_GEN_GCLK##n
```

mnkp commented:
This PR adds `soc/arm/atmel_sam0/common/power_samd2x.c` in the same commit. Likely a mistake.

sslupsky commented:
@mnkp The reason I included power_samd2x.c with this PR is it is required for exiting from sleep properly.  Specifically, sys_pm_power_state_exit_post_ops().

sslupsky commented:
@ioannisg I found an issue with the TICK_THRESHOLD when setting CYCLES_PER_TICK to something greater that 1024.  You can safely set CYCLES_PER_TICk to 2048 but anything more than that will be a problem because of the sync busy synchronizations.

carlescufi commented:
@sslupsky could you please rebase it?
@andyross and @mnkp any objections to merging this?

carlescufi commented:
@sslupsky do you mind if I rebase this myself?

mnkp commented:
@sslupsky could you please rebase this PR.

> The reason I included power_samd2x.c with this PR is it is required for exiting from sleep properly. Specifically, sys_pm_power_state_exit_post_ops().

Implementation of system power management on SAM0 family is a new, distinct feature. It should be added in a separate PR or at least in a separate commit.

sslupsky committed:
drivers: power_samd2x.c:  add sys power management

This commit enables system power management for the samd2x.

The samd2x rtc requires special attention when coming out of sleep.
The kernel does not like it when the RTC does not tick
after waking from sleep.  RCONT enables continuously syncing the
COUNT register but after sleep, we need to wait for a new read
request to complete.  However, the SYNCBUSY flag is not set
when RCONT is enabled so we cannot use that to do the sync.
So, we wait for the RTC COUNT register to begin ticking again by
reading the COUNT register.

We should wait a minimum of the TICK_THRESHOLD which is the amount
of time it takes to sync the register.

Signed-off-by: Steven Slupsky <sslupsky@gmail.com>

sslupsky commented:
@mnkp Ok, I created a separate commit for the sys power management.  I think it best to keep it connected here because the two items are interdependent.

sslupsky commented:
@carlescufi I rebased and pushed but I think you need to rebase against the latest and I have not yet migrated.  Please go ahead and rebase it.

carlescufi reviewed (approved):
Reviewed in detail offline

carlescufi commented:
@sslupsky I rebased this for you.
@nzmichaelh would you mind taking a look please?

nzmichaelh commented on drivers/timer/sam0_rtc_timer.c:
```diff
@@ -41,10 +45,13 @@
  * Due to the nature of clock synchronization, reading from or writing to some
- * RTC registers takes approximately six RTC_GCLK cycles. This constant defines
- * a safe threshold for the comparator.
+ * RTC registers takes approximately six RTC_GCLK cycles.
```
Update the comment ("six") to match.

sslupsky replied:
Do you mean to match the TICK_THRESHOLD?  The actual delay is 6 GCLK + 3 APB cycles.  Generally, APB >> GCLK.  However, if for some reason, they were the same, then the delay would be equivalent to 9 cycles.  I believe the original author assumed APB >> GCLK and "rounded up" to 7 GCLK cycles.

Do you want me to clarify that or change the threshold?

sslupsky replied:
/*
 * Due to the nature of clock synchronization, reading from or writing to some
 * RTC registers takes approximately six RTC_GCLK cycles.
 *
 * The actual delay is 6 GCLK + 3 APB cycles. Generally, APB >> GCLK.
 * However, if for some reason, the clocks were the same, then the delay would
 * be equivalent to 9 cycles.
 * 
 * We will assume APB >> GCLK and "round up" to 7 GCLK cycles.
 * The worst case update time after sleep requires 2 sync busy synchronizations.
 * Thus, the threshold is 7 x 2 = 14 cycles.
 *
 * This constant defines a safe threshold for the comparator.
 */

nzmichaelh commented on drivers/timer/sam0_rtc_timer.c:
```diff
@@ -102,10 +110,21 @@ static inline void rtc_sync(void)
+	u32_t ret;
+
+	rtc_sync();
+	if ((RTC0->READREQ.reg & RTC_READREQ_RCONT) == 0) {
```
Use the READREQ.bit.RCONT version instead.

sslupsky replied:
done

nzmichaelh commented on drivers/timer/sam0_rtc_timer.c:
```diff
@@ -102,10 +110,21 @@ static inline void rtc_sync(void)
+
+	rtc_sync();
+	if ((RTC0->READREQ.reg & RTC_READREQ_RCONT) == 0) {
+		RTC0->READREQ.reg |= RTC_READREQ_RCONT;
```
Ditto.  Optionally, put both line 117 and 118 into a single `|=`

sslupsky replied:
I do not think you can combine these into one operation.  The datasheet does appear to indicate that RCONT must be written before the RREQ bit.
_For the continuous read mode, RREQ bit is required to be set once the RCONT bit is set._

I think in my experimentation, I attempted to combine them.  If I recall, it does not work reliably if you do.

Overall, the entire RCONT functionality is poorly documented.

sslupsky replied:
I should also mention these changes only affect the SAM D.  I do not presently have SAM E hardware to do testing of the RCONT function for that device.  I reviewed the SAM E datasheet and believe the RCONT functionality may be different.
  
So, these changes are targeting the SAM D only at the moment.

nzmichaelh commented on drivers/timer/sam0_rtc_timer.c:
```diff
@@ -102,10 +110,21 @@ static inline void rtc_sync(void)
+		rtc_sync();
+	}
+
+	ret = RTC0->COUNT.reg;
```
Drop the assignment and do 'return RTC0->COUNT.reg'.

sslupsky replied:
The assignment aided greatly during debugging.  I can change this back.

nzmichaelh commented on drivers/timer/sam0_rtc_timer.c:
```diff
@@ -179,11 +199,11 @@ int z_clock_driver_init(struct device *device)
 {
 	ARG_UNUSED(device);
 
+	/* Set up bus clock and GCLK generator. */
```
This comment is no longer correct, as it's in front of the MCLK block which has no GCLK.

sslupsky replied:
I will clarify the comment.

nzmichaelh commented on soc/arm/atmel_sam0/common/power_samd2x.c:
```diff
@@ -0,0 +1,100 @@
+/*
```
This commit seems unrelated to the RTC fix.  Please split it out.

sslupsky replied:
These changes are related and required because a special synchronization is required when coming out of the STANDBY sleep state.  Specifically, _sys_pm_power_state_exit_post_ops() provides a busy wait until the RTC starts ticking again.

sslupsky commented:
Hold off merging this for a bit.  I have noticed an anomaly with the wake from sleep that needs further investigation.

carlescufi commented:
> Hold off merging this for a bit. I have noticed an anomaly with the wake from sleep that needs further investigation.

OK, let me know when it's ready.
